### PR TITLE
Backend refactor

### PR DESCRIPTION
**Pull request by robdockins**
This PR untangles the `Backend` class and it's implementation from the `Eval` module subtree.  This separates the concerns of different base-type value representation from the concerns of the evaluator proper and it's datatypes.

This PR consists almost entirely of file moves and module import changes, and other small code motion required to get things to compile again.  The largest change that isn't just a file move was the splitting of `Cryptol.Eval.SBV` into two parts, similar to how the other backends were already split.

This sets the stage for reuse of the Backend class in other settings (e.g., saw-core).

### CODE REVIEW

File: src/Cryptol/Backend/What4.hs [hunk 1]
```diff
@@ -28,17 +28,18 @@ import qualified GHC.Integer.GMP.Internals as Integer
...
-import Cryptol.Eval.Monad
+import Cryptol.Backend
+import Cryptol.Backend.Concrete( BV(..), ppBV )
+import Cryptol.Backend.FloatHelpers
+import Cryptol.Backend.Monad
    ( Eval(..), EvalError(..), Unsupported(..)
    , delayFill, blackhole, evalSpark
    )
+
+import Cryptol.Eval.Generic
```
Comment (robdockins) [inline src/Cryptol/Backend/What4.hs:42]: Still have some imports from `Eval` here

**Comment by robdockins** [comment]
At this point the `Backend` module subtree is nearly independent of Cryptol.  The things that still tether the backend modules to Cryptol are the `EvalError` datatype and the various `EvalOptions` settings.  It would be fairly easy to sever the options types, I think, but the EvalError type is baked into the evaluation monads in various places.

File: src/Cryptol/Backend/Monad.hs [hunk 2]
```diff
@@ -139,7 +139,7 @@ data ThunkState a
        --   a tight loop evaluating this thunk.  If the thunk is being evaluated
        --   by some other thread, the current thread will await it's completion.
```
Comment (brianhuffman) [inline src/Cryptol/Backend/Monad.hs:140]: You missed the same typo two lines earlier :)

Comment (robdockins) [reply src/Cryptol/Backend/Monad.hs:140]: c'est la vie

**Review by brianhuffman — APPROVED** [review]
This looks great; thanks for doing this! I'm excited about how much code reuse we should be able to get by using this part of the Cryptol API within saw-core.

**Pull request merged**